File: clavier_propre/controller.py
"""Contrôleur central : orchestre la bascule entre les trois cibles
(clavier physique, Microsoft Office, LibreOffice) et conserve l'état.
"""
from __future__ import annotations

import logging
from dataclasses import dataclass

from clavier_propre import keyboard, office_lo, office_ms
from clavier_propre.config import AppConfig

log = logging.getLogger(__name__)


@dataclass(frozen=True)
class ProtectionState:
    # ``True`` = protection active (suggestions et corrections DÉSACTIVÉS).
    active: bool
    keyboard_enabled: bool  # True = suggestions clavier actuellement actives
    word_managed: bool
    libreoffice_managed: bool


class ProtectionController:
    """Fait le pont entre la GUI/service et les modules de pilotage."""

    def __init__(self, config: AppConfig | None = None) -> None:
        self.config = config or AppConfig.load()

    def current_state(self) -> ProtectionState:
        kb = keyboard.get_state()
        return ProtectionState(
            active=self.config.protection_active,
            keyboard_enabled=kb.enabled,
            word_managed=self.config.manage_word,
            libreoffice_managed=self.config.manage_libreoffice,
        )

    def apply(self, active: bool) -> ProtectionState:
        """Active (``active=True``) ou désactive la protection globale."""
        self.config.protection_active = active
        self.config.save()

        # 1) Clavier physique : on désactive les suggestions quand la
        #    protection est active.
        keyboard.set_suggestions(enabled=not active)

        # 2) Microsoft Office (Word/Excel/PowerPoint).
        if self.config.manage_word:
            try:
                office_ms.apply_office_protection(active=active)
            except Exception as exc:  # pragma: no cover
                log.warning("Pilotage Microsoft Office échoué: %s", exc)

        # 3) LibreOffice.
        if self.config.manage_libreoffice:
            try:
                office_lo.apply_libreoffice_protection(active=active)
            except Exception as exc:  # pragma: no cover
                log.warning("Pilotage LibreOffice échoué: %s", exc)

        return self.current_state()

    def set_word_managed(self, managed: bool) -> None:
        self.config.manage_word = managed
        self.config.save()
        if self.config.protection_active:
            try:
                office_ms.apply_office_protection(active=managed)
            except Exception as exc:  # pragma: no cover
                log.warning("Pilotage Word échoué: %s", exc)

    def set_libreoffice_managed(self, managed: bool) -> None:
        self.config.manage_libreoffice = managed
        self.config.save()
        if self.config.protection_active:
            try:
                office_lo.apply_libreoffice_protection(active=managed)
            except Exception as exc:  # pragma: no cover
                log.warning("Pilotage LibreOffice échoué: %s", exc)

File: clavier-rs/src/controller.rs
//! Contrôleur central : orchestre la bascule entre les trois cibles.

use crate::config::AppConfig;
use crate::keyboard;
use crate::office_lo;
use crate::office_ms;
use std::sync::Mutex;

pub struct ProtectionController {
    pub config: Mutex<AppConfig>,
}

impl ProtectionController {
    pub fn new() -> Self {
        Self {
            config: Mutex::new(AppConfig::load()),
        }
    }

    /// Applique l'état de protection global.
    /// `active=true` -> suggestions et corrections DÉSACTIVÉES.
    pub fn apply(&self, active: bool) {
        {
            let mut cfg = self.config.lock().unwrap();
            cfg.protection_active = active;
            cfg.save();
        }
        // Clavier physique : on désactive les suggestions quand la protection
        // est active.
        keyboard::set_suggestions(!active);

        let cfg = self.config.lock().unwrap();
        if cfg.manage_word {
            office_ms::apply_office_protection(active);
        }
        if cfg.manage_libreoffice {
            office_lo::apply_libreoffice_protection(active);
        }
    }

    pub fn set_word_managed(&self, managed: bool) {
        let active = {
            let mut cfg = self.config.lock().unwrap();
            cfg.manage_word = managed;
            cfg.save();
            cfg.protection_active
        };
        if active {
            office_ms::apply_office_protection(managed);
        }
    }

    pub fn set_libreoffice_managed(&self, managed: bool) {
        let active = {
            let mut cfg = self.config.lock().unwrap();
            cfg.manage_libreoffice = managed;
            cfg.save();
            cfg.protection_active
        };
        if active {
            office_lo::apply_libreoffice_protection(managed);
        }
    }

    pub fn protection_active(&self) -> bool {
        self.config.lock().unwrap().protection_active
    }

    /// Watchdog : si la protection doit être active mais que les suggestions
    /// ont été réactivées, on ré-applique.
    pub fn watchdog(&self) {
        let active = self.protection_active();
        if active && keyboard::suggestions_enabled() {
            eprintln!("Watchdog : réactivation détectée, re-protection.");
            self.apply(true);
        }
    }
}

impl Default for ProtectionController {
    fn default() -> Self {
        Self::new()
    }
}
